engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateByWindowCumulative.java

```diff
@@ -0,0 +1,274 @@
+                                : context.timestampColumnSource.makeGetContext(context.workingChunkSize)) {
+            while (it.hasMore()) {
+                final RowSequence rs = it.getNextRowSequenceWithLength(context.workingChunkSize);
+                final int size = rs.intSize();
+                Arrays.fill(context.inputSourceChunkPopulated, false);
+
+                // create the timestamp chunk if needed
+                LongChunk<? extends Values> tsChunk = context.timestampColumnSource == null ? null
+                        : context.timestampColumnSource.getChunk(tsGetCtx, rs).asLongChunk();
+
+                for (int opIdx : context.dirtyOperatorIndices) {
+                    // prep the chunk array needed by the accumulate call
+                    final int[] srcIndices = operatorInputSourceSlots[opIdx];
+                    Chunk<? extends Values>[] chunkArr = new Chunk[srcIndices.length];
+                    for (int ii = 0; ii < srcIndices.length; ii++) {
+                        int srcIdx = srcIndices[ii];
+                        // chunk prep
+                        prepareValuesChunkForSource(context, srcIdx, rs);
+                        chunkArr[ii] = context.inputSourceChunks[srcIdx];
+                    }
+
+                    // make the specialized call for cumulative operators
+                    ((UpdateByCumulativeOperator.Context) context.opContext[opIdx]).accumulate(
+                            rs,
+                            chunkArr,
+                            tsChunk,
+                            size);
+                }
+            }
+        }
+
+        // call `finishUpdate()` function for each operator
+        for (int opIdx : context.dirtyOperatorIndices) {
+            operators[opIdx].finishUpdate(context.opContext[opIdx]);
+        }
+    }
+
+
+    /**
+     * Find the smallest valued key that participated in the upstream {@link TableUpdate}.
+     *
+     * @param added the added rows
+     * @param modified the modified rows
+     * @param removed the removed rows
+     * @param shifted the shifted rows
+     *
+     * @return the smallest key that participated in any part of the update.
+     */
+    private static long smallestAffectedKey(@NotNull final RowSet added,
+            @NotNull final RowSet modified,
+            @NotNull final RowSet removed,
+            @NotNull final RowSetShiftData shifted,
+            @NotNull final RowSet affectedIndex) {
+
+        long smallestModifiedKey = Long.MAX_VALUE;
+        if (removed.isNonempty()) {
+            smallestModifiedKey = removed.firstRowKey();
+        }
+
+        if (added.isNonempty()) {
+            smallestModifiedKey = Math.min(smallestModifiedKey, added.firstRowKey());
+        }
+
+        if (modified.isNonempty()) {
+            smallestModifiedKey = Math.min(smallestModifiedKey, modified.firstRowKey());
+        }
+
+        if (shifted.nonempty()) {
```

Comment:
We should get @abaranec to explain this to us. I maintain that shifts cannot affect a cumulative operator. Couldn't we just take `removed.firstRowKey()` and shift it, and then take the min of that, `modified.firstRowKey()`, and `added.firstRowKey()` to find out lower bound in post-shift space?

Reply:
Implemented a variation of this, shifting removed rows seems tricky since they aren't (apparently) represented in the `shifted()` ranges.